collections/src/main/scala/strawman/collection/immutable/LazyList.scala

```diff
@@ -434,84 +523,136 @@ object LazyList extends SeqFactory[LazyList] {
+  } else LazyList.Empty
 
   def empty[A]: LazyList[A] = Empty
+}
 
-  /** An infinite LazyList that repeatedly applies a given function to a start value.
-   *
-   *  @param start the start value of the LazyList
-   *  @param f     the function that's repeatedly applied
-   *  @return      the LazyList returning the infinite sequence of values `start, f(start), f(f(start)), ...`
-   */
-  def iterate[A](start: => A)(f: A => A): LazyList[A] = cons(start, iterate(f(start))(f))
+@deprecated("Use LazyList (which has a lazy head and tail) instead of Stream (which has a lazy tail only)", "2.13.0")
+sealed abstract class Stream[+A] extends LinearSeq[A] with LazyListOps[A, Stream, Stream[A]] {
+  def iterableFactory: LazyListFactory[Stream] = Stream
 
-  /**
-    * Create an infinite LazyList starting at `start` and incrementing by
-    * step `step`.
+  protected[this] def fromSpecificIterable(coll: collection.Iterable[A]): Stream[A] = fromIterable(coll)
+
+  protected[this] def newSpecificBuilder(): Builder[A, Stream[A]] =
+    IndexedSeq.newBuilder().mapResult(_.to(Stream))
+
+  protected[this] def cons[T](hd: => T, tl: => Stream[T]): Stream[T] = new Stream.Cons(hd, tl)
+
+  /** Apply the given function `f` to each element of this linear sequence
+    * (while respecting the order of the elements).
     *
-    * @param start the start value of the LazyList
-    * @param step the increment value of the LazyList
-    * @return the LazyList starting at value `start`.
+    *  @param f The treatment to apply to each element.
+    *  @note  Overridden here as final to trigger tail-call optimization, which
+    *  replaces 'this' with 'tail' at each iteration. This is absolutely
+    *  necessary for allowing the GC to collect the underlying LazyList as elements
+    *  are consumed.
+    *  @note  This function will force the realization of the entire LazyList
+    *  unless the `f` throws an exception.
     */
-  def from(start: Int, step: Int): LazyList[Int] =
-    cons(start, from(start + step, step))
+  @tailrec
+  override final def foreach[U](f: A => U): Unit = {
+    if (!this.isEmpty) {
+      f(head)
+      tail.foreach(f)
+    }
+  }
 
-  /**
-   * Create an infinite LazyList starting at `start` and incrementing by `1`.
-   *
-   * @param start the start value of the LazyList
-   * @return the LazyList starting at value `start`.
-   */
-  def from(start: Int): LazyList[Int] = from(start, 1)
+  /** LazyList specialization of foldLeft which allows GC to collect along the
+    * way.
+    *
+    * @tparam B The type of value being accumulated.
+    * @param z The initial value seeded into the function `op`.
+    * @param op The operation to perform on successive elements of the `LazyList`.
+    * @return The accumulated value from successive applications of `op`.
+    */
+  @tailrec
+  override final def foldLeft[B](z: B)(op: (B, A) => B): B = {
+    if (this.isEmpty) z
+    else tail.foldLeft(op(z, head))(op)
+  }
+}
```

Comment:
Why isn’t it possible to put `foreach` and `foldLeft` in `LazyListOps`?

Reply:
They can't be tail-recursive because the types of `this` and `tail` can be different.

Reply:
And implementing `foreach` with a `while` loop doesn't work, either, because you cannot drop the `this` reference manually.